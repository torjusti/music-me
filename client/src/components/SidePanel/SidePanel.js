import React, { Component } from 'react';
import Slider from 'rc-slider';
import { connect } from 'react-redux';
import 'rc-slider/assets/index.css';
import { Icon, Button } from 'semantic-ui-react';
import styles from './SidePanel.module.css';
import { setGenreSelected } from '../../features/genres/actions';
import sortBy from 'lodash/sortBy';
import {
  setRatingSelected,
  toggleRatingEnabled,
} from '../../features/rating/actions';

/**
 * The sidebar component, which handles the client-side part
 * of the majority of the filtering features.
 */
class SidePanel extends Component {
  state = {
    // The currently selected rating. This also exists
    // in Redux, however the value stored there does
    // not update until you drop the cursor, and thus, this
    // value is required.
    selectedRating: 1,
  };

  /**
   * Set the currently selected rating.
   */
  setValue = rating => {
    this.setState({ selectedRating: rating });
  };

  /**
   * Notify Redux about the updated rating when the cursor is released.
   */
  update = () => {
    this.props.setRatingSelected(this.state.selectedRating);
  };

  render() {
    const availableGenres = sortBy(this.props.genres.availableGenres, 'genre');

    return (
      <div className={styles.controls}>
        <h1 className="ui header">Filter menu</h1>

        <div className={styles.controlsCont}>
          <h3 className={styles.filterHeader}>Filter by rating</h3>

          <Button
            toggle
            active={this.props.rating.ratingEnabled}
            onClick={this.props.toggleRatingEnabled}
          >
            <Icon name="star" />
            Enable filter by rating
          </Button>

          <Slider
            min={1}
            max={5}
            className={styles.slider}
            value={this.state.selectedRating}
            onChange={this.setValue}
            onAfterChange={this.update}
            disabled={!this.props.rating.ratingEnabled}
            marks={{
              1: '1',
              2: '2',
              3: '3',
              4: '4',
              5: '5',
            }}
          />
        </div>

        <div className={styles.controlsCont}>
          <h3 className={styles.filterHeader}>Filter by genre</h3>

          <ul className={styles.noDecoration}>
            {availableGenres.map(elem => {
              const selected = this.props.genres.selectedGenres.includes(
                elem.genre,
              );

              return (
                <li key={elem.genre}>
                  <div className="ui checkbox">
                    <input
                      type="checkbox"
                      id={elem.genre}
                      checked={selected}
                      onChange={() =>
                        this.props.setGenreSelected(elem.genre, !selected)
                      }
                    />

                    <label htmlFor={elem.genre}>{elem.genre}</label>
                  </div>
                </li>
              );
            })}
          </ul>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  genres: state.genres,
  rating: state.rating,
});

const mapDispatchToProps = {
  setGenreSelected,
  setRatingSelected,
  toggleRatingEnabled,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SidePanel);
