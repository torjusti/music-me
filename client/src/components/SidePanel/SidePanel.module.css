.sliderCont {
  flex: 1;
}

.text {
  background-color: #ff0;
}
