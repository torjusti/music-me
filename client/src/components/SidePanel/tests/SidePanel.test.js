import React from 'react';
import { shallow } from 'enzyme';
import { SidePanel, mapStateToProps, mapDispatchToProps } from '../SidePanel';
import configureMockStore from 'redux-mock-store'

describe('<SidePanel />', () => {

  // Create the mock store
  const mockStore = configureMockStore();

  let component, store, initialState;

  beforeEach(()=>{
    initialState = {
      genres: {
        availableGenres: ["test1", "test2"],
        selectedGenres: ["test1"],
      },
      selectedRating: 3,
    };

    store = mockStore(initialState);
    // Shallow render the container passing in the mock store
    component = shallow(
      <SidePanel store={store} />
    );
  });

  it('should render SidePanel correctly', () => {
    expect(component.length).toEqual(1);
  });


  it('sidepanel snapshot should be equal', () => {
    expect(component).toMatchSnapshot();
  });

  it('', () => {
    expect(mapStateToProps(initialState).selectedRating).toEqual(3);
  });

  /*
  it('should update rating only by sidepanels update()', () => {

    const newRating = 4;
    const instance = wrapper.instance();
    instance.setValue(newRating);
    instance.update();
  })


  it('', () => {
    // test that the component events dispatch the expected actions
    wrapper.simulate('rollDice');

    const actions = store.getActions();
    expect(actions).toEqual([ { type: 'ROLL_DICE' } ]);
  });
  */

});
