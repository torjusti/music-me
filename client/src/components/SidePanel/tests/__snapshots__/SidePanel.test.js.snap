// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidePanel /> sidepanel snapshot should be equal 1`] = `undefined`;
